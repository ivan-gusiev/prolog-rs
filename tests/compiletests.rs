extern crate insta;
extern crate parameterized;
extern crate prolog_rs;

#[cfg(test)]
mod compiletests {
    use insta::assert_display_snapshot;
    use parameterized::parameterized;
    use prolog_rs::{
        compile::{compile_program, compile_query},
        lang::parse_term,
        symbol::SymbolTable,
        util::{case, writeout_sym},
    };

    #[parameterized(input = {
        "p(Z, h(Z,W), f(W))",
        "D",
        "f(X, g(X,a))",
        "f(b, Y)",
        "a"
    })]
    fn test_query_compile(input: &str) {
        let mut symbol_table = SymbolTable::new();
        let query = parse_term(input, &mut symbol_table).unwrap();
        let instructions = compile_query(query);

        assert_display_snapshot!(case(input, writeout_sym(&instructions, &symbol_table)));
    }

    #[parameterized(input = {
        "p(Z, h(Z,W), f(W))",
        "D",
        "f(X, g(X,a))",
        "f(b, Y)",
        "p(f(X), h(Y, f(a)), Y)",
        "a"
    })]
    fn test_program_compile(input: &str) {
        let mut symbol_table = SymbolTable::new();
        let query = parse_term(input, &mut symbol_table).unwrap();
        let instructions = compile_program(query);

        assert_display_snapshot!(case(input, writeout_sym(&instructions, &symbol_table)));
    }
}
