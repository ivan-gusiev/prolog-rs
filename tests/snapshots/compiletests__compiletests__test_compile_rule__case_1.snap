---
source: tests/compiletests.rs
expression: "case(input, collapse(result))"
---
f(X) :- p(X, r(Z, Y)), h(l(p(A, X), p(B, X))).
-----
000	allocate 1
001	get_variable Y1, X1	// Y1=X
002	put_value Y1, X1	// Y1=X
003	put_structure r/2, X2
004	set_variable X4
005	set_variable X5
006	call @120
007	put_structure p/2, X2
008	set_variable X4
009	set_value Y1	// Y1=X
010	put_structure p/2, X3
011	set_variable X6
012	set_value Y1	// Y1=X
013	put_structure l/2, X1
014	set_value X2
015	set_value X3
016	call @130
017	deallocate

